docs/modules/ROOT/pages/proxies.adoc

```diff
@@ -0,0 +1,368 @@
+In order to invoke `\\__l1_default__`, the original function call must include the library function `send_message_to_l1`.
+See Cairo's https://www.cairo-lang.org/docs/hello_starknet/l1l2.html[Interacting with L1 contracts] for more information.
+
+Since this proxy is designed to work both as an https://eips.ethereum.org/EIPS/eip-1822[UUPS-flavored upgrade proxy] as well as a non-upgradeable proxy, it does not know how to handle its own state.
+Therefore it requires the implementation contract class to be declared beforehand, so its class hash can be passed to the Proxy on construction time.
+
+When interacting with the contract, function calls should be sent by the user to the proxy.
+The proxy's fallback function redirects the function call to the implementation contract to execute.
+
+=== Implementation contract
+
+The implementation contract, also known as the logic contract, receives the redirected function calls from the proxy contract.
+The implementation contract should follow the link:../docs/Extensibility.md#the-pattern[Extensibility pattern] and import directly from the link:../src/openzeppelin/upgrades/library.cairo[Proxy library].
+
+The implementation contract should:
+
+* import `Proxy` namespace
+* initialize the proxy immediately after contract deployment with `Proxy.initializer`.
+
+If the implementation is upgradeable, it should:
+
+* include a method to upgrade the implementation (i.e.
+`upgrade`)
+* use access control to protect the contract's upgradeability.
+
+The implementation contract should NOT:
+
+* be deployed like a regular contract.
+Instead, the implementation contract should be declared (which creates a `DeclaredClass` containing its hash and abi)
+* set its initial state with a traditional constructor (decorated with `@constructor`).
+Instead, use an initializer method that invokes the Proxy `constructor`.
+
+NOTE: The Proxy `constructor` includes a check the ensures the initializer can only be called once;
+however, `_set_implementation` does not include this check.
+It's up to the developers to protect their implementation contract's upgradeability with access controls such as <<assert_only_admin,`assert_only_admin`>>.
+
+For a full implementation contract example, please see:
+
+* link:../tests/mocks/proxiable_implementation.cairo[Proxiable implementation]
+
+== Upgrades library API
+
+=== Methods
+
+[,cairo]
+----
+func initializer(proxy_admin: felt):
+end
+
+func assert_only_admin():
+end
+
+func get_implementation_hash() -> (implementation: felt):
+end
+
+func get_admin() -> (admin: felt):
+end
+
+func _set_admin(new_admin: felt):
+end
+
+func _set_implementation_hash(new_implementation: felt):
+end
+----
+
+==== `initializer`
+
+Initializes the proxy contract with an initial implementation.
+
+Parameters:
+
+[,cairo]
+----
+proxy_admin: felt
+----
+
+Returns:
+
+None.
+
+==== `assert_only_admin`
+
+Reverts if called by any account other than the admin.
+
+Parameters:
+
+None.
+
+Returns:
+
+None.
```

Comment:
I don't think the `None.` looks great in the API because it's indistinguishable from `Parameters:` and `Returns:`. One solution could be to insert it into a code block. I think it looks better, but I don't know. Maybe there's something else? This is how it looks:
![Screen Shot 2022-07-15 at 1 52 31 PM](https://user-images.githubusercontent.com/69282788/179286252-cc4b491e-bc33-4314-af40-b5dc34e669c7.png)

Reply:
I think if the "None" was moved onto the same line as "Parameters:" or "Returns:" that'd be better.

Reply:
Agreed. I think that's easier to follow.

Reply:
This is how it actually looks without making any changes thanks to the docsite's styling

<img width="778" alt="image" src="https://user-images.githubusercontent.com/1703831/180504965-fa16d715-76c1-4639-ba41-5c9097633406.png">